### PR TITLE
Font Library: Convert heading text to heading elements and group fonts as a list

### PR DESCRIPTION
**Pull request by t-hamano**
Related to #55220
Part of #58403
See this comment: https://github.com/WordPress/gutenberg/issues/58403#issuecomment-1932358501

## What?

This PR achieves the following two things in the font library modal.

- Convert heading text to heading elements
- Group fonts as a list

font-list

## Why?

~~User-installed fonts don't have the title, so Users don't know what that font list means.~~ Also, since the font list is actually just a list of button elements, users using screen readers cannot understand how many fonts there are.

## How?

Group the font list as follows.

```html
<h2>XXX Fonts</h2>
<ul>
	<li>
		<button>...</button>
	</li>
	<li>
		<button>...</button>
	</li>
</ul>
```

> [!NOTE]
> ~~By using `fieldset` and `legend`, we should be able to properly associate the heading with the content. But maybe these elements should be inside the `form` element. I would appreciate it if you could give me advice on whether this implementation violates a11y.~~

### Testing Instructions for Keyboard

- Launch a screen reader such as NVDA or VoiceOver.
- Open the font library modal just by using the keyboard.
- When you focus on the beginning of the font list: You should hear something like this: `Library  property page  list  with X items  {fontName}  graphic    X/X variants active  button  

## Screenshots or screencast 

https://github.com/WordPress/gutenberg/assets/54422211/919415e2-4dc4-43b8-b11b-9da9efe76b1e

### CODE REVIEW

File: packages/edit-site/src/components/global-styles/font-library-modal/fonts-grid.js
Comment (afercia) [inline packages/edit-site/src/components/global-styles/font-library-modal/fonts-grid.js]: I'm not sure using a H2 heading _within a fieldset legend_ makes much sense. It is true that semantically a fieldset elemetn has a native role of `group` but it's a bit overkill. I think headeings and lists already provide the necessary semantics and information to users. I'd ratehr remove the fieldset and legend entirely.

File: packages/edit-site/src/components/global-styles/font-library-modal/fonts-grid.js
Comment (afercia) [inline packages/edit-site/src/components/global-styles/font-library-modal/fonts-grid.js]: Good idea using a list. This will help assistive technology to _count and announce_ the amount of items. Reminder that Safari and VoiceOver do not announce styled lists as list. The workaround is to add a `role="list"` even though that's technically redundant.

**Review by afercia — CHANGES_REQUESTED** [review]

**Comment by afercia** [comment]
I'm all for adding headings and using an unordered list. Headings clearly identify sections of content, both visually and semantically. Screenshot fo the headings hierarchy reported by teh headingsMap browser extensiont:

Screenshot 2024-02-08 at 15 59 51

File: packages/edit-site/src/components/global-styles/font-library-modal/fonts-grid.js
Comment (t-hamano) [reply packages/edit-site/src/components/global-styles/font-library-modal/fonts-grid.js]: Which of the following do you think is ideal for this context?

```html
	<ul>
		<li><button>...</button></li>
		<li><button>...</button></li>
	</ul>
```

```html
	<ul role="list">
		<li role="listitem"><button>...</button></li>
		<li role="listitem"><button>...</button></li>
	</ul>
```

```html
	<div class="list">
		<div class="listitem"><button>...</button></div>
		<div class="listitem"><button>...</button></div>
	</div>
```

File: packages/edit-site/src/components/global-styles/font-library-modal/fonts-grid.js
Comment (t-hamano) [reply packages/edit-site/src/components/global-styles/font-library-modal/fonts-grid.js]: Fixed with [bdbe0bc](https://github.com/WordPress/gutenberg/pull/58834/commits/bdbe0bcd2704f3a69e24d054d18fe1f3eb968f2c)

File: packages/edit-site/src/components/global-styles/font-library-modal/fonts-grid.js
Comment (afercia) [reply packages/edit-site/src/components/global-styles/font-library-modal/fonts-grid.js]: role="listitem" on the items isn't necessary.
It's Safari that doesn't expose a `<ul>` as a role `list` when it's styled. Long time ago I had a twitters thread with folks from the Safari team and they called this a `feature`. Actually, it's a non-standard and non-compliant implementation. In other words, it's a Safari bug.

Screenshots from the Safari dev tools, without and with role=llist:

safari no ul role

safari ul role list

Comment (t-hamano) [reply packages/edit-site/src/components/global-styles/font-library-modal/fonts-grid.js]: Fixed with [3f5630d](https://github.com/WordPress/gutenberg/pull/58834/commits/3f5630db0a9c5081ee123e90622c3b748a2e7089)

Thank you for explaining in detail. I looked over the Gutenberg project again and found that `role=list` was indeed assigned to the `ul` element. However, this attribute is detected by eslint as an error, so it seems like we need to use `eslint-disable/enable`.

Comment (afercia) [reply packages/edit-site/src/components/global-styles/font-library-modal/fonts-grid.js]: Oh yes it could also be reported as warning by some automated accessibility checkers. Technically not a real error, it's just re-stating the native role of a `<ul>`, which is technically redundant.

Edit: Thanks Apple for this.

File: packages/edit-site/src/components/global-styles/font-library-modal/style.scss
Comment (colorful-tones) [inline packages/edit-site/src/components/global-styles/font-library-modal/style.scss]: I wonder if we could be more explicit and only override what we need here? I'm not a huge fan of shorthand for [these reasons](https://www.tempertemper.net/blog/why-i-prefer-not-to-use-css-shorthand). This is a nitpick and should not hold up this being merged.

```
margin-bottom: 0;
margin-top: 0;
```

File: packages/edit-site/src/components/global-styles/font-library-modal/style.scss
Comment (colorful-tones) [inline packages/edit-site/src/components/global-styles/font-library-modal/style.scss]: This could be 

```
margin-bottom: 0;
margin-top: -1px; /* To collapse the margin with the previous element */
```

File: packages/edit-site/src/components/global-styles/font-library-modal/style.scss
Comment (colorful-tones) [inline packages/edit-site/src/components/global-styles/font-library-modal/style.scss]: This could be

```
margin-bottom: 0;
margin-top: 0;
```

**Review by colorful-tones — APPROVED** [review]
This is a great adjustment, and thanks for diving in @t-hamano !

I left a few CSS nit-picks, but please do not let it stop from merging. I would ultimately think we want to be consistent with whatever Gutenberg's CSS standards are, and I'm rusty on that.

File: packages/edit-site/src/components/global-styles/font-library-modal/fonts-grid.js
Comment (colorful-tones) [reply packages/edit-site/src/components/global-styles/font-library-modal/fonts-grid.js]: I was totally unaware of this Safari bug and thanks for sharing @afercia

**Comment by t-hamano** [comment]
Thank you everyone for your advice. This PR will greatly conflict with  #58794, which replaces infinite scroll with pagination. Both are accessibility improvements, but  #58794 is probably a higher priority. Therefore, I plan to resolve the conflict with this PR after  #58794 is merged.

**Comment by t-hamano** [comment]
Since #58794 was merged, changes were made again to match the latest trunk. The following points, which are the objectives of this PR, should still be maintained.

- Add a title to the user font
- User font title and theme font title are h2 elements
- Properly mark up the font list with the ul element
- @colorful-tones 's feedback regarding margins has also been applied.

**Comment by t-hamano** [comment]
I added #55220, which this PR solves, to the WP6.5 project board and also gave this PR a backport label.

WP6.5 will probably see a lot of backporting regarding the font library. In order to avoid conflicts during the backport process, I think we need to backport as much as possible when there are changes to the code related to the font library.

**Comment by t-hamano** [comment]
Updated again: The font library was significantly refactored in trunk, so I resolved the conflict again and updated all PR descriptions.

**Comment by annezazu** [comment]
@WordPress/gutenberg-design can you all review this? I can see reviews were requested but a review hasn't quite been done. This overlaps nicely with what we've seen with grouping templates  https://github.com/WordPress/gutenberg/pull/57711 so I imagine this is another overlapping area.

**Comment by t-hamano** [comment]
Note: [A discussion](https://wordpress.slack.com/archives/C02RQBWTW/p1708987810984969) started in Slack about whether the Font Library should be shipped to the core. If the Font Library will not ship to WP6.5, we may want to remove the backport label.

**Comment by jasmussen** [comment]
I'm personally okay to move forward with this as an immediate fix, but I think there's a larger challenge with how the font library is coming together at the moment that needs clarification. 

It's easy to think that this new Font Library should be the only source of managing everything related to fonts, both across files and font stacks and theme fonts. But it can never be that one canonical source; some typographic styles will have to live in global styles, etc. So unless we get clarity about the role of the font library, this will become increasingly messy, as [also discussed here](https://github.com/WordPress/gutenberg/pull/59294#issuecomment-1961780883) related to where you manage font stacks. 

A cleaner delineation, which is also how this was designed, is to have the Font Library behave similarly to the Media Library: it's a place where you add, remove, manage files, and perhaps a little metadata. This delineation immediately provides clear answers to questions like these: where should font stacks be managed? Where should theme vs. user fonts live? Global styles, because the font library is only ever fonts you install yourself.

**Comment by t-hamano** [comment]
Remove the backport label and also exclude it from the project board. See [this comment](https://github.com/WordPress/gutenberg/issues/55220#issuecomment-1966410076).

**Comment by t-hamano** [comment]
@jasmussen Thank you for your opinion.

I agree that there is a larger challenge to be solved when it comes to the font library. After several rounds of conflict resolution regarding this PR, we finally settled on just improving accessibility, and as a result, no changes were made to the design. I should have told you that design feedback is no longer required 😅

**Comment by t-hamano** [comment]
@jasmussen @colorful-tones @afercia Could you review this PR? The content of the PR has not changed, but the conflict has been resolved again 🙏

File: packages/edit-site/src/components/global-styles/font-library-modal/style.scss
Comment (jasmussen) [inline packages/edit-site/src/components/global-styles/font-library-modal/style.scss:69]: Not a blocker from me, but in general it would be nice to move away from custom CSS in favor of the base components with their default styles. I assume this would require a larger refactor?

**Review by jasmussen — APPROVED** [review]

File: packages/edit-site/src/components/global-styles/font-library-modal/style.scss
Comment (t-hamano) [reply packages/edit-site/src/components/global-styles/font-library-modal/style.scss:69]: Similar to these styles is the label for the `BaseControl` component. However, we cannot use the label as a heading element.

image

Personally, I think it's fine as is for now.

Comment (jasmussen) [reply packages/edit-site/src/components/global-styles/font-library-modal/style.scss:69]: It would be nice to have a heading style for that appearance that can be used in more cases. (Agree, not a blocker.)

**Pull request merged**
